packages/docs/docs/guides/architecture.md

```diff
@@ -0,0 +1,168 @@
+
+:::note What are roundtrips?
+A roundtrip is when your client needs to make a connection to your backend. Depending on the quality of the user's network connection, this can have a big negative impact on performance. As a result, many performance strategies involve minimizing roundtrips.
+
+Any clientside app can have roundtrips (calling out to APIs, etc). With AI apps, we add a new type of roundtrip: calling out to a model provider (e.g. OpenAI).
+
+The amount of roundtrips in your logic depends on how you structure your AI.JSX program. A program with many sequential calls out to a model will have more roundtrips than one that does a single shot. (Of course, unlike traditional API calls, model calls are so slow that the client/server latency is a less important contributor to the overall performance profile.)
+
+For more details, see [Performance](./performance.md).
+:::
+
+:::tip Where can AI.JSX run?
+In addition to the client, AI.JSX can run in serverless/edge functions, traditional standalone servers, or any other NodeJS process.
+:::
+
+## Run Entirely on the Client
+
+With a pure client-side approach, you run your UI and AI.JSX logic on the client. However, you still need to call out to external APIs for hosted services like model providers (e.g. GPT-4 from OpenAI) or databases (e.g. Pinecone).
+
+Pros:
+
+- Simple to implement
+- The serialization boundary is at the API layer. The only way to access these APIs is through a serialization boundary, so your architecture isn't adding any new boundaries.
+
+Cons:
+
+- Your API keys are exposed to the client.
+- Performance will suffer if you need to do many roundtrips
+
+```mermaid
+sequenceDiagram
+    box rgb(200, 200, 255) Client
+    participant UI
+    participant AI.JSX
+    end
+    box rgb(200, 255, 200) Third-Party API
+    participant APIs
+    end
+    AI.JSX->>APIs: Model calls (e.g. to OpenAI)
+    AI.JSX->>APIs: Other API calls (e.g. to Pinecone)
+```
+
+## UI + AI.JSX on the client; API calls on the server
+
+This is just like the [Run Entirely on the Client](#run-entirely-on-the-client) pattern, except you add a proxy for API calls. You'd traditionally set this proxy up as a serverless or edge function. The proxy stores your API keys, keeping them safe from the client.
+
+Pros:
+
+- There's still only one serialization boundary.
+- API keys are protected from the client.
+
+Cons:
+
+- Compared to pure client, it's a little more complicated because you need to set up a proxy.
+
+```mermaid
+sequenceDiagram
+    box rgb(200, 200, 255) Client
+    participant UI
+    participant AI.JSX
+    end
+    box rgb(200, 255, 200) API Proxy
+    participant Your Proxy
+    end
+    AI.JSX->>APIs: Model calls (e.g. to OpenAI)
+    AI.JSX->>APIs: Other API calls (e.g. to Pinecone)
+    box rgb(200, 255, 200) Third-Party API
+    participant APIs
+    end
+```
+
+## UI on the client; AI.JSX on the server
+
+In this approach, your AI.JSX logic runs entirely in your backend environment.
```

Comment:
does next.js/RSC enter into the equation here?

Reply:
I pushed a change to this doc and others that hopefully addresses this.